### PR TITLE
llvm: Update and refactor the UDF compiler

### PR DESCRIPTION
**Pull request by jvesely**
Extend test parametrization to cover more corner cases.
Add support for expressions of non-boolean type in if condition.
Add support for non-boolean values in boolean expressions.
Add support for subtraction and power operators.
Add support for unary add and not.
Fix return statements to return from the function instead of just updating the return value.
Allocate variable space for variable storage, the type is derived from the first occurrence of the variable in code.
Fix the NaN behavior of Python and numpy max builtins.
Refactor expression generation to not use stack space for constructing temporary values (with the exception of subscript).

### CODE REVIEW

**Review by SamKG — APPROVED** [review]

File: psyneulink/core/llvm/codegen.py [hunk 1]
```diff
@@ -455,12 +455,35 @@ def visit_BinOp(self, node):
...
+            # Python's 'and' takes the boolean value of the first operand and returns
+            # RHS if bool(LHS) else LHS
+            cond = helpers.convert_type(builder, x, self.ctx.bool_ty)
+            return builder.select(cond, y, x)
+
+        return _and
+
+    def visit_Or(self, node):
+        def _or(builder, x, y):
+            # Python's 'and' takes the boolean value of the first operand and returns
```
[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/codegen.py]: updated. thanks.

File: psyneulink/core/llvm/codegen.py [hunk 2]
```diff
@@ -160,6 +160,14 @@ def _add(builder, x, y):
 
         return _add
 
+    def visit_Sub(self, node):
+        def _add(builder, x, y):
```
[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/codegen.py]: renamed. thanks

File: tests/functions/test_user_defined_func.py [hunk 3]
```diff
@@ -406,34 +406,50 @@ def myFunction(x,t0=0.48):
...
+    elif dtype == "BOOL_VAR":
         def myFunction(variable):
             var = True
             return 1.0
-    elif dtype == "TUPLE":
+    elif dtype == "TUPLE_VAR":
         def myFunction(variable):
             var = (1, 2, 3, 4)
             return var
+    elif dtype == "SCALAR_LIT":
```
[the remark this answers is not in the document]
Comment (jvesely) [reply tests/functions/test_user_defined_func.py:454]: you're right it tests mostly return variants now. renamed. thanks.
